File: backend/app/providers/mock.py
"""Deterministic mock FareProvider. No network calls, no cost -- used
automatically whenever SERPAPI_API_KEY is unset (FARE_PROVIDER=auto) and
always in tests. Same (origin, destination, dates, currency) always
produces the same options, forever, so pruning/ranking tests can assert
on exact outcomes and "deals" can be deliberately planted and found.

Price model (all multipliers deterministic given the seeded RNG below):
    price = route_baseline[destination]
          * origin_factor[origin]      -- simulates secondary-airport savings
          * seasonal(depart_date)
          * weekday_factor(depart_date)
          * trip_length_factor(nights)
          * deal_factor (7% chance of a planted deal)
          * noise (0.93-1.08)
          * adults                     -- linear, no group discount (MVP simplification)
"""

from __future__ import annotations

import datetime as dt
import hashlib
import json
import math
import random

from app.config import SEED_DIR
from app.providers.base import FareLeg, FareOption, FareProvider, FareQuery, FareSlice, MultiCityLeg

# Independent from services/indicative.py's ONE_WAY_FACTOR (that one
# approximates a cold-start baseline before any real observations exist;
# this one IS the simulated market price) -- importing one from the other
# would also be circular, since indicative.py already imports ORIGIN_FACTOR
# from this module.
ONE_WAY_PRICE_FACTOR = 0.6

# Relative pricing pull per Toronto-region origin airport. Deliberately
# gives BUF/YHM a strong pull below YYZ so the nearby-airport savings
# logic (spec §31: min_saving_per_person) has real deals to find in tests.
ORIGIN_FACTOR = {
    "YYZ": 1.00,
    "YTZ": 1.05,
    "YHM": 0.88,
    "YKF": 0.95,
    "BUF": 0.75,
}

# Plausible single connecting hub per destination, for a one-stop option.
# Not real routing data -- just enough structure for the mock landscape
# to exercise stops/layover logic.
CONNECTING_HUB = {
    "HND": "ICN", "NRT": "ICN", "KIX": "ICN", "ITM": "HND", "UKB": "KIX",
    "ICN": "HND", "TPE": "HKG", "HKG": "TPE", "SIN": "HKG", "BKK": "HKG",
    "IST": "AMS", "DOH": "IST", "DXB": "IST",
    "LHR": "AMS", "CDG": "AMS", "AMS": "CDG", "FRA": "AMS", "LIS": "CDG", "OPO": "LIS",
}

CARRIER_POOL = ["AC", "NH", "OZ", "KE", "CI", "CX", "SQ", "TG", "TK", "QR", "EK", "BA", "AF", "KL", "LH", "TP"]

_AIRPORT_COORDS: dict[str, tuple[float, float]] | None = None


def _airport_coords() -> dict[str, tuple[float, float]]:
    global _AIRPORT_COORDS
    if _AIRPORT_COORDS is None:
        with (SEED_DIR / "airports.json").open("r", encoding="utf-8") as f:
            rows = json.load(f)
        _AIRPORT_COORDS = {r["iata"]: (r["lat"], r["lon"]) for r in rows}
    return _AIRPORT_COORDS


def _haversine_km(a: str, b: str) -> float:
    coords = _airport_coords()
    lat1, lon1 = coords.get(a, (43.6777, -79.6248))  # fall back to YYZ if unknown
    lat2, lon2 = coords.get(b, (43.6777, -79.6248))
    r = 6371.0
    p1, p2 = math.radians(lat1), math.radians(lat2)
    dphi = math.radians(lat2 - lat1)
    dlambda = math.radians(lon2 - lon1)
    x = math.sin(dphi / 2) ** 2 + math.cos(p1) * math.cos(p2) * math.sin(dlambda / 2) ** 2
    return 2 * r * math.asin(math.sqrt(x))


def _leg_duration_min(a: str, b: str) -> int:
    km = _haversine_km(a, b)
    cruise_hours = km / 830.0  # rough average incl. routing inefficiency
    return round(cruise_hours * 60) + 40  # + taxi/climb/descent overhead


_BASELINES: dict[str, float] | None = None


def _baseline(destination: str) -> float:
    global _BASELINES
    if _BASELINES is None:
        with (SEED_DIR / "route_baselines.json").open("r", encoding="utf-8") as f:
            _BASELINES = json.load(f)["toronto"]
    return _BASELINES.get(destination, 1100.0)


def _seeded_rng(query: FareQuery) -> random.Random:
    material = f"{query.origin}|{query.destination}|{query.depart_date.isoformat()}|{query.return_date.isoformat()}|{query.currency}"
    seed_int = int(hashlib.sha256(material.encode()).hexdigest()[:16], 16)
    return random.Random(seed_int)


def _seasonal_factor(d: dt.date) -> float:
    day_of_year = d.timetuple().tm_yday
    return 1.0 + 0.12 * math.sin(2 * math.pi * day_of_year / 365.0)


def _weekday_factor(d: dt.date) -> float:
    # Mon=0 .. Sun=6. Cheaper mid-week departures, pricier Fri/Sun.
    return {0: 1.00, 1: 0.95, 2: 0.94, 3: 0.97, 4: 1.08, 5: 1.02, 6: 1.08}[d.weekday()]


def _trip_length_factor(nights: int) -> float:
    if nights < 7:
        return 1.15
    if nights > 21:
        return 1.10
    return 1.0


def _price_for(query: FareQuery, rng: random.Random, *, is_onestop: bool) -> float:
    base = _baseline(query.destination)
    nights = (query.return_date - query.depart_date).days
    price = (
        base
        * ORIGIN_FACTOR.get(query.origin, 1.0)
        * _seasonal_factor(query.depart_date)
        * _weekday_factor(query.depart_date)
        * _trip_length_factor(nights)
    )
    if is_onestop:
        price *= rng.uniform(0.85, 0.94)  # connecting itineraries tend cheaper
    if rng.random() < 0.07:
        price *= 0.74  # planted deal
    price *= rng.uniform(0.93, 1.08)  # noise
    price *= query.adults
    return round(price, 2)


def _price_for_one_way(query: FareQuery, rng: random.Random, *, is_onestop: bool) -> float:
    """Deliberately does not delegate to _price_for: that formula's nights
    factor is meaningless here (there is no return leg), and reusing it
    with nights=0 would wrongly apply the <7-night short-trip surcharge.
    """
    base = _baseline(query.destination)
    price = (
        base
        * ONE_WAY_PRICE_FACTOR
        * ORIGIN_FACTOR.get(query.origin, 1.0)
        * _seasonal_factor(query.depart_date)
        * _weekday_factor(query.depart_date)
    )
    if is_onestop:
        price *= rng.uniform(0.85, 0.94)
    if rng.random() < 0.07:
        price *= 0.74  # planted deal
    price *= rng.uniform(0.93, 1.08)  # noise
    price *= query.adults
    return round(price, 2)


def _make_option(
    query: FareQuery, rng: random.Random, *, hub: str | None, price: float
) -> FareOption:
    dep_hour = rng.randint(6, 22)
    dep_dt = dt.datetime.combine(query.depart_date, dt.time(hour=dep_hour))
    carrier = rng.choice(CARRIER_POOL)

    if hub is None:
        duration = _leg_duration_min(query.origin, query.destination)
        arr_dt = dep_dt + dt.timedelta(minutes=duration)
        leg = FareLeg(
            from_iata=query.origin, to_iata=query.destination,
            dep_time=dep_dt, arr_time=arr_dt, carrier=carrier,
            flight_number=f"{carrier}{rng.randint(100, 999)}", duration_min=duration,
        )
        return FareOption(
            price=price, currency=query.currency, outbound_legs=(leg,), layovers=(),
            total_duration_min=duration, stops=0, carriers=(carrier,),
            inbound_detail="full", raw={"mock": True, "template": "nonstop"},
        )

    leg1_min = _leg_duration_min(query.origin, hub)
    layover_min = rng.randint(70, 260)
    leg2_min = _leg_duration_min(hub, query.destination)
    leg1_arr = dep_dt + dt.timedelta(minutes=leg1_min)
    leg2_dep = leg1_arr + dt.timedelta(minutes=layover_min)
    leg2_arr = leg2_dep + dt.timedelta(minutes=leg2_min)
    carrier2 = rng.choice(CARRIER_POOL)
    legs = (
        FareLeg(query.origin, hub, dep_dt, leg1_arr, carrier, f"{carrier}{rng.randint(100, 999)}", leg1_min),
        FareLeg(hub, query.destination, leg2_dep, leg2_arr, carrier2, f"{carrier2}{rng.randint(100, 999)}", leg2_min),
    )
    total = leg1_min + layover_min + leg2_min
    return FareOption(
        price=price, currency=query.currency, outbound_legs=legs, layovers=((hub, layover_min),),
        total_duration_min=total, stops=1, carriers=tuple(dict.fromkeys((carrier, carrier2))),
        inbound_detail="full", raw={"mock": True, "template": "onestop", "hub": hub},
    )


class MockFareProvider:
    name = "mock"

    async def search_round_trip(self, query: FareQuery) -> list[FareOption]:
        rng = _seeded_rng(query)
        options: list[FareOption] = []

        options.append(_make_option(query, rng, hub=None, price=_price_for(query, rng, is_onestop=False)))

        hub = CONNECTING_HUB.get(query.destination)
        if hub and hub != query.origin:
            options.append(_make_option(query, rng, hub=hub, price=_price_for(query, rng, is_onestop=True)))

        if rng.random() < 0.5:
            options.append(_make_option(query, rng, hub=None, price=_price_for(query, rng, is_onestop=False)))

        if hub and rng.random() < 0.4:
            options.append(_make_option(query, rng, hub=hub, price=_price_for(query, rng, is_onestop=True)))

        if query.max_stops is not None:
            options = [o for o in options if o.stops <= query.max_stops]

        options.sort(key=lambda o: o.price)
        return options

    async def search_one_way(self, query: FareQuery) -> list[FareOption]:
        rng = _seeded_rng(query)
        options: list[FareOption] = []

        options.append(_make_option(query, rng, hub=None, price=_price_for_one_way(query, rng, is_onestop=False)))

        hub = CONNECTING_HUB.get(query.destination)
        if hub and hub != query.origin:
            options.append(_make_option(query, rng, hub=hub, price=_price_for_one_way(query, rng, is_onestop=True)))

        if query.max_stops is not None:
            options = [o for o in options if o.stops <= query.max_stops]

        options.sort(key=lambda o: o.price)
        return options

    async def search_multi_city(
        self,
        legs: list[MultiCityLeg],
        *,
        adults: int = 1,
        currency: str = "CAD",
        max_stops: int | None = None,
    ) -> list[FareOption]:
        """Returns exactly one combined itinerary (sum of per-leg prices),
        not multiple variants -- unlike search_round_trip/search_one_way,
        there's no small set of "the" alternatives to enumerate for an
        N-leg chain without combinatorial blowup, and this is a mock
        provider, not a search-quality benchmark. Per-leg nonstop-vs-hub
        routing is still deterministic (seeded on that leg's own
        origin/destination/date), so repeated searches are stable.
        """
        slices: list[FareSlice] = []
        all_legs: list[FareLeg] = []
        all_layovers: list[tuple[str, int]] = []
        all_carriers: list[str] = []
        total_price = 0.0
        total_duration = 0
        total_stops = 0

        for leg in legs:
            leg_query = FareQuery(
                origin=leg.origin,
                destination=leg.destination,
                depart_date=leg.date,
                return_date=leg.date,
                adults=adults,
                currency=currency,
                max_stops=max_stops,
                trip_type="one_way",
            )
            rng = _seeded_rng(leg_query)
            hub = CONNECTING_HUB.get(leg.destination)
            use_hub = bool(hub and hub != leg.origin and rng.random() < 0.35)
            price = _price_for_one_way(leg_query, rng, is_onestop=use_hub)
            option = _make_option(leg_query, rng, hub=hub if use_hub else None, price=price)

            total_price += option.price
            all_legs.extend(option.outbound_legs)
            all_layovers.extend(option.layovers)
            all_carriers.extend(option.carriers)
            total_duration += option.total_duration_min
            total_stops += option.stops
            slices.append(
                FareSlice(
                    legs=option.outbound_legs,
                    layovers=option.layovers,
                    stops=option.stops,
                    duration_min=option.total_duration_min,
                )
            )

        combined = FareOption(
            price=round(total_price, 2),
            currency=currency,
            outbound_legs=tuple(all_legs),
            layovers=tuple(all_layovers),
            total_duration_min=total_duration,
            stops=total_stops,
            carriers=tuple(dict.fromkeys(all_carriers)),
            inbound_detail="full",
            raw={"mock": True, "template": "multi_city"},
            slices=tuple(slices),
        )
        return [combined]


mock_provider: FareProvider = MockFareProvider()

File: frontend/src/lib/engine/mock.ts
// Deterministic in-browser mock provider. Zero network, zero cost. Two
// jobs: (1) the "demo" mode -- the API-key gate accepts the literal key
// "demo" so anyone can exercise the full UI without a SerpApi account or
// spending quota; (2) engine tests. Loosely mirrors the Python mock
// provider's market model (baseline x origin factor x seasonal x weekday
// x deal x noise, plausible hub connections) but does NOT reproduce its
// exact numbers -- Python's Mersenne-Twister semantics aren't worth
// porting; determinism per query is what matters, not cross-language
// numeric parity. Same (query) always -> same options, forever.

import { estimateDistanceKm } from "./mockGeo";
import { ROUTE_BASELINES } from "./seed";
import type { EngineProvider, FareLeg, FareOption, FareQuery, FareSlice } from "./types";

export const MOCK_API_KEY = "demo";

const ORIGIN_FACTOR: Record<string, number> = { YYZ: 1.0, YTZ: 1.05, YHM: 0.88, YKF: 0.95, BUF: 0.75 };

const CONNECTING_HUB: Record<string, string> = {
  HND: "ICN", NRT: "ICN", KIX: "ICN", ITM: "HND", UKB: "KIX",
  ICN: "HND", TPE: "HKG", HKG: "TPE", SIN: "HKG", BKK: "HKG",
  IST: "AMS", DOH: "IST", DXB: "IST",
  LHR: "AMS", CDG: "AMS", AMS: "CDG", FRA: "AMS", LIS: "CDG", OPO: "LIS",
};

const CARRIER_POOL = ["AC", "NH", "OZ", "KE", "CI", "CX", "SQ", "TG", "TK", "QR", "EK", "BA", "AF", "KL", "LH", "TP"];

// --- Seeded PRNG: FNV-1a string hash -> mulberry32 ---

function fnv1a(material: string): number {
  let hash = 0x811c9dc5;
  for (let i = 0; i < material.length; i++) {
    hash ^= material.charCodeAt(i);
    hash = Math.imul(hash, 0x01000193);
  }
  return hash >>> 0;
}

class Rng {
  private state: number;

  constructor(material: string) {
    this.state = fnv1a(material);
  }

  random(): number {
    this.state = (this.state + 0x6d2b79f5) >>> 0;
    let t = this.state;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  }

  uniform(lo: number, hi: number): number {
    return lo + (hi - lo) * this.random();
  }

  randint(lo: number, hi: number): number {
    return lo + Math.floor(this.random() * (hi - lo + 1));
  }

  choice<T>(arr: T[]): T {
    return arr[Math.floor(this.random() * arr.length)];
  }
}

function legDurationMin(from: string, to: string): number {
  const km = estimateDistanceKm(from, to);
  return Math.round((km / 830) * 60) + 40;
}

function baseline(destination: string): number {
  return ROUTE_BASELINES.toronto?.[destination] ?? 1100;
}

function seasonalFactor(dateIso: string): number {
  const d = new Date(`${dateIso}T00:00:00Z`);
  const start = Date.UTC(d.getUTCFullYear(), 0, 0);
  const dayOfYear = Math.floor((d.getTime() - start) / 86_400_000);
  return 1.0 + 0.12 * Math.sin((2 * Math.PI * dayOfYear) / 365);
}

function weekdayFactor(dateIso: string): number {
  // getUTCDay: Sun=0..Sat=6 -> cheaper mid-week, pricier Fri/Sun.
  return [1.08, 1.0, 0.95, 0.94, 0.97, 1.08, 1.02][new Date(`${dateIso}T00:00:00Z`).getUTCDay()];
}

function tripLengthFactor(nights: number): number {
  if (nights < 7) return 1.15;
  if (nights > 21) return 1.1;
  return 1.0;
}

function priceFor(query: FareQuery, rng: Rng, isOnestop: boolean): number {
  const nights = Math.round((Date.parse(query.returnDate) - Date.parse(query.departDate)) / 86_400_000);
  let price =
    baseline(query.destination) *
    (ORIGIN_FACTOR[query.origin] ?? 1.0) *
    seasonalFactor(query.departDate) *
    weekdayFactor(query.departDate) *
    // One-way has no nights-based pricing; ~60% of the round-trip level.
    (query.tripType === "one_way" ? 0.6 : tripLengthFactor(nights));
  if (isOnestop) price *= rng.uniform(0.85, 0.94);
  if (rng.random() < 0.07) price *= 0.74; // planted deal
  price *= rng.uniform(0.93, 1.08);
  price *= query.adults;
  return Math.round(price * 100) / 100;
}

function isoAt(dateIso: string, hour: number, minuteOffset = 0): string {
  const t = new Date(`${dateIso}T00:00:00Z`);
  t.setUTCMinutes(hour * 60 + minuteOffset);
  return t.toISOString().slice(0, 16);
}

function makeLegs(
  query: Pick<FareQuery, "origin" | "destination" | "departDate">,
  rng: Rng,
  hub: string | null,
): { legs: FareLeg[]; layovers: [string, number][] } {
  const depHour = rng.randint(6, 22);
  const carrier = rng.choice(CARRIER_POOL);

  if (hub === null) {
    const duration = legDurationMin(query.origin, query.destination);
    return {
      legs: [
        {
          from_iata: query.origin,
          to_iata: query.destination,
          dep_time: isoAt(query.departDate, depHour),
          arr_time: isoAt(query.departDate, depHour, duration),
          carrier,
          flight_number: `${carrier}${rng.randint(100, 999)}`,
          duration_min: duration,
        },
      ],
      layovers: [],
    };
  }

  const leg1Min = legDurationMin(query.origin, hub);
  const layoverMin = rng.randint(70, 260);
  const leg2Min = legDurationMin(hub, query.destination);
  const carrier2 = rng.choice(CARRIER_POOL);
  return {
    legs: [
      {
        from_iata: query.origin,
        to_iata: hub,
        dep_time: isoAt(query.departDate, depHour),
        arr_time: isoAt(query.departDate, depHour, leg1Min),
        carrier,
        flight_number: `${carrier}${rng.randint(100, 999)}`,
        duration_min: leg1Min,
      },
      {
        from_iata: hub,
        to_iata: query.destination,
        dep_time: isoAt(query.departDate, depHour, leg1Min + layoverMin),
        arr_time: isoAt(query.departDate, depHour, leg1Min + layoverMin + leg2Min),
        carrier: carrier2,
        flight_number: `${carrier2}${rng.randint(100, 999)}`,
        duration_min: leg2Min,
      },
    ],
    layovers: [[hub, layoverMin]],
  };
}

function makeOption(query: FareQuery, rng: Rng, hub: string | null, price: number): FareOption {
  const { legs, layovers } = makeLegs(query, rng, hub);
  const total = legs.reduce((s, l) => s + l.duration_min, 0) + layovers.reduce((s, [, m]) => s + m, 0);
  return {
    price,
    currency: query.currency,
    outbound_legs: legs,
    layovers,
    total_duration_min: total,
    stops: layovers.length,
    carriers: [...new Set(legs.map((l) => l.carrier))],
    inbound_detail: "full",
    slices: [],
  };
}

function searchDeterministic(query: FareQuery): FareOption[] {
  const rng = new Rng(
    `${query.origin}|${query.destination}|${query.departDate}|${query.returnDate}|${query.currency}|${query.tripType}`,
  );
  const options: FareOption[] = [];

  options.push(makeOption(query, rng, null, priceFor(query, rng, false)));

  const hub = CONNECTING_HUB[query.destination];
  if (hub && hub !== query.origin) {
    options.push(makeOption(query, rng, hub, priceFor(query, rng, true)));
  }
  if (rng.random() < 0.5) {
    options.push(makeOption(query, rng, null, priceFor(query, rng, false)));
  }
  if (hub && rng.random() < 0.4) {
    options.push(makeOption(query, rng, hub, priceFor(query, rng, true)));
  }

  const filtered = query.maxStops === null ? options : options.filter((o) => o.stops <= query.maxStops!);
  filtered.sort((a, b) => a.price - b.price);
  return filtered;
}

export const mockProvider: EngineProvider = {
  name: "mock",

  async searchRoundTrip(query) {
    return searchDeterministic(query);
  },

  async searchOneWay(query) {
    return searchDeterministic(query);
  },

  async searchMultiCity(legs, opts) {
    // One deterministic option: each requested leg becomes a slice
    // (nonstop or one-stop via the hub table), priced like a one-way and
    // summed. Inter-slice gaps never appear in layovers -- matching the
    // real provider's semantics (see serpapi.ts buildMultiCityOption).
    const slices: FareSlice[] = [];
    let price = 0;
    for (const leg of legs) {
      const query: FareQuery = {
        origin: leg.origin,
        destination: leg.destination,
        departDate: leg.date,
        returnDate: leg.date,
        adults: opts.adults,
        currency: opts.currency,
        maxStops: opts.maxStops,
        tripType: "one_way",
      };
      const rng = new Rng(`mc|${leg.origin}|${leg.destination}|${leg.date}|${opts.currency}`);
      const hub = rng.random() < 0.4 ? (CONNECTING_HUB[leg.destination] ?? null) : null;
      const usableHub = hub && hub !== leg.origin ? hub : null;
      const { legs: sliceLegs, layovers } = makeLegs(query, rng, usableHub);
      const duration = sliceLegs.reduce((s, l) => s + l.duration_min, 0) + layovers.reduce((s, [, m]) => s + m, 0);
      const slice: FareSlice = { legs: sliceLegs, layovers, stops: sliceLegs.length - 1, duration_min: duration };
      if (opts.maxStops !== null && slice.stops > opts.maxStops) return [];
      slices.push(slice);
      price += priceFor(query, rng, usableHub !== null);
    }
    const allLegs = slices.flatMap((sl) => sl.legs);
    return [
      {
        price: Math.round(price * 100) / 100,
        currency: opts.currency,
        outbound_legs: allLegs,
        layovers: slices.flatMap((sl) => sl.layovers),
        total_duration_min: slices.reduce((s, sl) => s + sl.duration_min, 0),
        stops: slices.reduce((s, sl) => s + sl.stops, 0),
        carriers: [...new Set(allLegs.map((l) => l.carrier))],
        inbound_detail: "full",
        slices,
      },
    ];
  },
};
